File: tests/SnapBrief.Windows.Tests/ClipboardEchoDetectorTests.cs
namespace SnapBrief.Windows.Tests;

public sealed class ClipboardEchoDetectorTests
{
    private const string PromptText = "Снимок A.\r\n  A1: тест1";

    [Fact]
    public void ReceiverEchoWithImagePlaceholder_IsDetected()
    {
        var snapshot = TextOnly("[Image #2]Снимок A.\r\n  A1: тест1");

        Assert.True(ClipboardEchoDetector.IsReceiverEcho(snapshot, PromptText));
    }

    [Fact]
    public void ReceiverEchoWithExtraWhitespaceAndCrlf_IsDetected()
    {
        var snapshot = TextOnly("  Снимок   A.\n\n  A1:   тест1  \r\n");

        Assert.True(ClipboardEchoDetector.IsReceiverEcho(snapshot, PromptText));
    }

    [Fact]
    public void UnrelatedForeignText_IsNotAnEcho()
    {
        var snapshot = TextOnly("Совсем другой текст, скопированный пользователем");

        Assert.False(ClipboardEchoDetector.IsReceiverEcho(snapshot, PromptText));
    }

    [Fact]
    public void ForeignTextWithFiles_IsNotAnEcho()
    {
        var data = new Dictionary<string, object>(StringComparer.Ordinal)
        {
            ["FileDrop"] = new[] { @"C:\shots\0.png" },
            ["UnicodeText"] = "Совсем другой текст"
        };
        var snapshot = new ClipboardSnapshot(99, data, true);

        Assert.False(ClipboardEchoDetector.IsReceiverEcho(snapshot, PromptText));
    }

    [Fact]
    public void ClipboardWithImage_IsNotAnEcho()
    {
        var data = new Dictionary<string, object>(StringComparer.Ordinal)
        {
            ["Bitmap"] = new object(),
            ["UnicodeText"] = PromptText
        };
        var snapshot = new ClipboardSnapshot(99, data, true);

        Assert.False(ClipboardEchoDetector.IsReceiverEcho(snapshot, PromptText));
    }

    private static ClipboardSnapshot TextOnly(string text)
    {
        var data = new Dictionary<string, object>(StringComparer.Ordinal) { ["UnicodeText"] = text };
        return new ClipboardSnapshot(99, data, true);
    }
}

File: macos/Tests/SnapikCoreTests/Transport/ClipboardEchoDetectorTests.swift
// XCTest port of tests/Snapik.Windows.Tests/ClipboardEchoDetectorTests.cs (5 tests),
// SPEC-DELTA-2 Part 1 §5.

import XCTest
@testable import SnapikCore

final class ClipboardEchoDetectorTests: XCTestCase {
    private static let prompt = "Снимок A.\r\n  A1: тест1"

    // ReceiverEchoWithImagePlaceholder_IsDetected
    func testReceiverEchoWithImagePlaceholderIsDetected() {
        let snapshot = Self.textOnly("[Image #2]Снимок A.\r\n  A1: тест1")
        XCTAssertTrue(ClipboardEchoDetector.isReceiverEcho(snapshot, promptText: Self.prompt))
    }

    // ReceiverEchoWithExtraWhitespaceAndCrlf_IsDetected
    func testReceiverEchoWithExtraWhitespaceAndCrlfIsDetected() {
        let snapshot = Self.textOnly("  Снимок   A.\n\n  A1:   тест1  \r\n")
        XCTAssertTrue(ClipboardEchoDetector.isReceiverEcho(snapshot, promptText: Self.prompt))
    }

    // UnrelatedForeignText_IsNotAnEcho
    func testUnrelatedForeignTextIsNotAnEcho() {
        let snapshot = Self.textOnly("Совершенно другой текст")
        XCTAssertFalse(ClipboardEchoDetector.isReceiverEcho(snapshot, promptText: Self.prompt))
    }

    // ForeignTextWithFiles_IsNotAnEcho
    func testForeignTextWithFilesIsNotAnEcho() {
        let snapshot = ClipboardSnapshot(
            sequence: 1, hasText: true, text: Self.prompt, filePaths: ["/tmp/a.png"], hasImage: false)
        XCTAssertFalse(ClipboardEchoDetector.isReceiverEcho(snapshot, promptText: Self.prompt))
    }

    // ClipboardWithImage_IsNotAnEcho
    func testClipboardWithImageIsNotAnEcho() {
        let snapshot = ClipboardSnapshot(sequence: 1, hasText: true, text: Self.prompt, filePaths: [], hasImage: true)
        XCTAssertFalse(ClipboardEchoDetector.isReceiverEcho(snapshot, promptText: Self.prompt))
    }

    // Empty prompt text is never an echo (guard in `isReceiverEcho`'s first line).
    func testEmptyPromptIsNeverAnEcho() {
        let snapshot = Self.textOnly(Self.prompt)
        XCTAssertFalse(ClipboardEchoDetector.isReceiverEcho(snapshot, promptText: ""))
    }

    private static func textOnly(_ text: String) -> ClipboardSnapshot {
        ClipboardSnapshot(sequence: 1, hasText: true, text: text, filePaths: [], hasImage: false)
    }
}
